### PR TITLE
Import and export Redshift tables (fixes #26)

### LINKED ISSUE #26: redshift

### PR DESCRIPTION
**Pull request by emk**
To do:

- [x] Spike and set up test server.
- [x] Write end-to-end test.
- [x] Implement Redshift import.
- [x] Implement Redshift export.
- [x] Reduce the excessive number of parameters in the `Locator` API, because the old version made @seamusabshere sad and this version is worse.
- [ ] ~~Move all code shared between two drivers into an appropriate `drivers/*_shared` directory.~~ We can do this another time. It's pretty minor.
- [x] Fix https://github.com/faradayio/dbcrossbar/issues/40

## Notes on the `Locator` API refactoring and `Features`

The problem: We support N driver features and M drivers, but every feature interacts with every driver, giving us O(N*M) code complexity. This also means that adding a new driver requires O(N) effort (to at least print an error if a feature isn't supported) and adding a new feature requires O(M) effort, to examine each driver. This also means that `local_data`, `write_local_data` and `write_remote_data` have long and constantly-growning parameter lists.

The solution here has several parts:

1. Each driver must explicity declare the features it supports, using a
   set of bit flags in a `Features` struct. See https://docs.rs/bitflags.
2. Driver arguments are grouped into `SharedArguments`,
   `SourceArguments` and `DestinationArguments`, vastly simplifying
   parameter lists.
3. In order to use an `*Arguments` struct, it _must_ be verified against
   a `Features` list, so that any unsupported features can be rejected
   automatically. We enforce mandatory verification using a neat Rust
   trick known as "type states": http://cliffle.com/blog/rust-typestate/.
   This allows us to keep track--at compile time!--of which arguments
   have been verified and which haven't.

The net result of all this is that we could now add a feature like `--select`, and we would only need to change those drivers which we want to support `--select`. All other drivers will automatically print an error.

### CODE REVIEW

**Comment by emk** [comment]
This PR requires review by either @seamusabshere or @n1ywb. Once it is reviewed, we can ship support for Redshift and for PostgreSQL schemas. Please don't hesitate to ask questions. Thank you!

File: dbcrossbarlib/src/drivers/redshift/local_data.rs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    shared_args: SharedArguments<Unverified>,
+    source_args: SourceArguments<Unverified>,
+) -> Result<Option<BoxStream<CsvStream>>> {
+    // Build a temporary location.
+    let shared_args_v = shared_args.clone().verify(RedshiftLocator::features())?;
+    let s3_temp = find_s3_temp_dir(shared_args_v.temporary_storage())?;
+    let s3_dest_args = DestinationArguments::for_temporary();
+    let s3_source_args = SourceArguments::for_temporary();
+
+    // Extract from BigQuery to gs://.
```
Comment (prithajnath) [inline dbcrossbarlib/src/drivers/redshift/local_data.rs]: Should be Redshift

File: dbcrossbarlib/src/drivers/redshift/write_local_data.rs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    shared_args: SharedArguments<Unverified>,
+    dest_args: DestinationArguments<Unverified>,
+) -> Result<BoxStream<BoxFuture<()>>> {
+    // Build a temporary location.
+    let shared_args_v = shared_args.clone().verify(RedshiftLocator::features())?;
+    let s3_temp = find_s3_temp_dir(shared_args_v.temporary_storage())?;
+    let s3_dest_args = DestinationArguments::for_temporary();
+    let s3_source_args = SourceArguments::for_temporary();
+
+    // Copy to a temporary gs:// location.
```
Comment (prithajnath) [inline dbcrossbarlib/src/drivers/redshift/write_local_data.rs]:
```suggestion
    // Copy to a temporary s3:// location.
```

File: dbcrossbarlib/src/drivers/redshift/write_local_data.rs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    let to_temp_ctx = ctx.child(o!("to_temp" => s3_temp.to_string()));
+    let result_stream = s3_temp
+        .write_local_data(to_temp_ctx, data, shared_args.clone(), s3_dest_args)
+        .await?;
+
+    // Wait for all gs:// uploads to finish with controllable parallelism.
```
Comment (prithajnath) [inline dbcrossbarlib/src/drivers/redshift/write_local_data.rs]:
```suggestion
    // Wait for all s3:// uploads to finish with controllable parallelism.
```

File: dbcrossbarlib/src/drivers/redshift/write_local_data.rs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    //
+    // TODO: This duplicates our top-level `cp` code and we need to implement
+    // the same rules for picking a good argument to `consume_with_parallelism`
+    // and not just hard code our parallelism.
+    result_stream.consume_with_parallelism(4).await?;
+
+    // Load from gs:// to BigQuery.
```
Comment (prithajnath) [inline dbcrossbarlib/src/drivers/redshift/write_local_data.rs]:
```suggestion
    // Load from s3:// to Redshift.
```

File: dbcrossbarlib/src/drivers/redshift/write_local_data.rs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    dest.write_remote_data(
+        from_temp_ctx,
+        Box::new(s3_temp),
+        shared_args,
+        s3_source_args,
+        dest_args,
+    )
+    .await?;
+
+    // We don't need any parallelism after the BigQuery step, so just return
```
Comment (prithajnath) [inline dbcrossbarlib/src/drivers/redshift/write_local_data.rs]:
```suggestion
    // We don't need any parallelism after the Redshift step, so just return
```

File: dbcrossbarlib/src/drivers/redshift/local_data.rs (hunk 1 above)
Comment (emk) [inline dbcrossbarlib/src/drivers/redshift/local_data.rs]:
```suggestion
    // Extract from Redshift to s3://.
```

**Pull request merged**